From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Hygon <hygon806@gmail.com>
Date: Mon, 22 Aug 2022 22:53:52 +0200
Subject: [PATCH] Add defade maven repository


diff --git a/build-logic/src/main/kotlin/minestom.publishing-conventions.gradle.kts b/build-logic/src/main/kotlin/minestom.publishing-conventions.gradle.kts
index b63a95323b4e20af1c36f29f24c623d978016696..a9518601c3775e9a3654f410b7f5eecef54774c3 100644
--- a/build-logic/src/main/kotlin/minestom.publishing-conventions.gradle.kts
+++ b/build-logic/src/main/kotlin/minestom.publishing-conventions.gradle.kts
@@ -1,32 +1,18 @@
 plugins {
-    id("net.kyori.indra")
-    id("net.kyori.indra.publishing")
-    id("net.kyori.indra.publishing.sonatype")
+    id("maven-publish")
 }
 
-indra {
-    javaVersions {
-        target(17)
-        testWith(17)
+configure<PublishingExtension> {
+    publications.create<MavenPublication>("maven") {
+        from(components["java"])
     }
+}
 
-    github("Minestom", "Minestom") {
-        ci(true)
-    }
-    apache2License()
-
-    configurePublications {
-        pom {
-            developers {
-                developer {
-                    id.set("TheMode")
-                    name.set("TheMode")
-                }
-                developer {
-                    id.set("jglrxavpok")
-                    name.set("jglrxavpok")
-                }
-            }
+publishing {
+    repositories {
+        maven("https://repo.defade.net/defade") {
+            name = "defade"
+            credentials(PasswordCredentials::class)
         }
     }
 }
\ No newline at end of file
diff --git a/build.gradle.kts b/build.gradle.kts
index 5c8299fcc971fb481b428e22efdf1e92051f4b3c..39200e231bb32f3dcd595de3ed55ddbcc3e3aa99 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -5,9 +5,8 @@ plugins {
 }
 
 allprojects {
-    group = "net.minestom.server"
-    version = "1.0"
-    description = "Lightweight and multi-threaded Minecraft server implementation"
+    group = "net.defade"
+    version = "1.19.2-R1.0-SNAPSHOT"
 }
 
 sourceSets {
diff --git a/settings.gradle.kts b/settings.gradle.kts
index 2230bb02892d10d6756722105337ae6434cc8a35..a2d24392f10dc181932f4ae9e6699ed10e4ff3c1 100644
--- a/settings.gradle.kts
+++ b/settings.gradle.kts
@@ -16,7 +16,7 @@ pluginManagement {
     includeBuild("build-logic")
 }
 
-rootProject.name = "Minestom"
+rootProject.name = "Yokura"
 include("code-generators")
 include("jmh-benchmarks")
 include("jcstress-tests")
diff --git a/testing/build.gradle.kts b/testing/build.gradle.kts
index 4ea0e6bc68bea05e0345261a861fc9fcae9461fe..7a520c78af35311bdaeb37405c7b23ee3a0b4ad0 100644
--- a/testing/build.gradle.kts
+++ b/testing/build.gradle.kts
@@ -20,6 +20,10 @@ version = "1.0"
 
 repositories {
     mavenCentral()
+    maven("https://repo.defade.net/defade") {
+        name = "defade"
+        credentials(PasswordCredentials::class)
+    }
     maven(url = "https://jitpack.io")
 }
 
