From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Hygon <hygon806@gmail.com>
Date: Fri, 26 Aug 2022 01:22:59 +0200
Subject: [PATCH] Store player permissions to DB


diff --git a/src/main/java/net/minestom/server/entity/Player.java b/src/main/java/net/minestom/server/entity/Player.java
index 4b910b23b7166f744406abb08462ffa2336d79bd..dfa601bc1573f7adad4108a9ae5e1cd384a25c10 100644
--- a/src/main/java/net/minestom/server/entity/Player.java
+++ b/src/main/java/net/minestom/server/entity/Player.java
@@ -61,6 +61,7 @@ import net.minestom.server.network.packet.server.play.*;
 import net.minestom.server.network.player.GameProfile;
 import net.minestom.server.network.player.PlayerConnection;
 import net.minestom.server.network.player.PlayerSocketConnection;
+import net.minestom.server.permission.Permission;
 import net.minestom.server.recipe.Recipe;
 import net.minestom.server.recipe.RecipeManager;
 import net.minestom.server.resourcepack.ResourcePack;
@@ -93,7 +94,10 @@ import org.jetbrains.annotations.NotNull;
 import org.jetbrains.annotations.Nullable;
 import org.jglrxavpok.hephaistos.nbt.NBT;
 import org.jglrxavpok.hephaistos.nbt.NBTCompound;
+import org.jglrxavpok.hephaistos.nbt.NBTException;
+import org.jglrxavpok.hephaistos.parser.SNBTParser;
 
+import java.io.StringReader;
 import java.nio.charset.StandardCharsets;
 import java.time.Duration;
 import java.util.*;
@@ -252,7 +256,8 @@ public class Player extends LivingEntity implements CommandSender, Localizable,
                     .append("endermites", MinecraftServer.getBismuth().getNetworkInfos().getDefaultEndermites())
                     .append("crystals", MinecraftServer.getBismuth().getNetworkInfos().getDefaultCrystals())
                     .append("level", 0)
-                    .append("xp", 0);
+                    .append("xp", 0)
+                    .append("permissions", new ArrayList<>());
 
             MongoUtils.insertDocument("players", playerData);
         }
@@ -261,6 +266,19 @@ public class Player extends LivingEntity implements CommandSender, Localizable,
         crystals = playerData.getDouble("crystals");
         networkLevel = playerData.getInteger("level");
         networkXp = playerData.getInteger("xp");
+
+        for(Document permissions : playerData.getList("permissions", Document.class)) {
+            NBTCompound nbt = null;
+            if(permissions.getString("nbt") != null) {
+                try {
+                    nbt = (NBTCompound) new SNBTParser(new StringReader(permissions.getString("nbt"))).parse();
+                } catch (NBTException e) {
+                    MinecraftServer.getExceptionManager().handleException(e);
+                }
+            }
+
+            super.addPermission(new Permission(permissions.getString("permission"), nbt));
+        }
     }
 
     public void addEndermites(double endermites) {
@@ -334,6 +352,40 @@ public class Player extends LivingEntity implements CommandSender, Localizable,
         MongoUtils.executeAsync(() -> MongoUtils.updateValue("players", getUuid().toString(), "xp", getNetworkXp()));
     }
 
+    @Override
+    public void addPermission(@NotNull Permission permission) {
+        super.addPermission(permission);
+
+        MongoUtils.executeAsync(() -> {
+            Document permissionDocument = new Document("permission", permission.getPermissionName());
+            if(permission.getNBTData() != null) {
+                permissionDocument.append("nbt", permission.getNBTData().toSNBT());
+            }
+
+            MongoUtils.addToList("players", getUuid().toString(), "permissions", permissionDocument);
+        });
+    }
+
+    @Override
+    public void removePermission(@NotNull Permission permission) {
+        super.removePermission(permission);
+        MongoUtils.executeAsync(() -> {
+            Document filter = new Document("_id", getUuid().toString());
+            Document update = new Document("$pull", new Document("permissions", new Document("permission", permission.getPermissionName())));
+            MinecraftServer.getMongoDatabase().getCollection("players").updateOne(filter, update);
+        });
+    }
+
+    @Override
+    public void removePermission(@NotNull String permissionName) {
+        super.removePermission(permissionName);
+        MongoUtils.executeAsync(() -> {
+            Document filter = new Document("_id", getUuid().toString());
+            Document update = new Document("$pull", new Document("permissions", new Document("permission", permissionName)));
+            MinecraftServer.getMongoDatabase().getCollection("players").updateOne(filter, update);
+        });
+    }
+
     /**
      * Used when the player is created.
      * Init the player and spawn him.
